app/src/server/utils/parseTags.ts

```diff
@@ -1,6 +1,8 @@
 import { z } from "zod";
 
-export const parseTags = (tags: unknown) => {
+export function parseTags(tags: unknown): Record<string, string>;
+export function parseTags(tags: unknown, keepNulls: true): Record<string, string | null>;
+export function parseTags(tags: unknown, keepNulls?: true) {
   if (typeof tags !== "object") {
```

Comment:
Seems like some of these checks could be removed/simplified by starting with `const parsedTags = z.record(z.union([z.string(), z.number(), z.boolean(), z.null()]))` at the top here?

Reply:
I made the decision not to use zod for some of these checks because I'm not a big fan of the error messages it produces. I regret that we automated a large part of our dataset entry upload parsing pipeline with zod, since I think it's a lot harder to figure out what's actually going on for a first-time user.

Reply:
Fair, for certain kinds of errors we can definitely give more guidance manually than you'd get out of Zod. Although Zod actually lets you set custom error messages when a property fails validation, which might get us a lot of the way there while keeping it simplicity. https://github.com/colinhacks/zod/blob/master/ERROR_HANDLING.md#schema-bound-error-map